### PR TITLE
Jax operator and jax example

### PR DESCRIPTION
**Pull request by Ezra-H**

### CODE REVIEW

File: distml/operator/jax_operator.py [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+                                  "your model, optimizer, and criterion.")
+
+    def register(self,
+                 *,
+                 model,
+                 optimizer,
+                 criterion,
+                 jit_mode=False):
+        """Register a few critical information about the model to operator."""
+        self.criterion = criterion
```
Comment (zhisbug) [inline distml/operator/jax_operator.py:72]: @Ezra-H : please add a sanity check here to make sure `model`, `optimizer` and `critierion` are valida JAX types. See TorchTrainingOperator for examples.

File: distml/operator/jax_operator.py [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+
+         This function shall be instantiated in framework-specific operator
+         implementations.
+         """
+        self.opt_state = model[0]
+        self.init_fun = model[1]
+        self.predict_fun = model[2]
+
+    def _register_optimizer(self, optimizer):
+        self.opt_init = optimizer[0]
```
Comment (zhisbug) [inline distml/operator/jax_operator.py:87]: same for optimizer and its states...

File: distml/operator/jax_operator.py [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+        self._train_loader = train_loader
+        self._validation_loader = validation_loader
+
+    def _get_train_loader(self):
+        return self._train_loader
+
+    def _get_validation_loader(self):
+        return self._validation_loader
+
+    def loss_func(self, params, batch):
```
Comment (zhisbug) [inline distml/operator/jax_operator.py:101]: where are `params` for here?  please add docstring for it below. Also might need to check whether params are valid inputs.

File: distml/operator/jax_operator.py [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+                raise TypeError(msg.format(subtree, new_subtree))
+
+        self.opt_state = OptimizerState(new_states_flat, tree, new_subtrees)
+
+    def reset_optimizer_for_params(self, params):
+        keys, params = unzip2(sorted(params.items(), key=lambda d: int(d[0])))
+        self.tree = tree_structure(params)
+        self.opt_state = self.opt_init(params)
+
+    def ones(self, shape, cpu=True):
```
Comment (zhisbug) [inline distml/operator/jax_operator.py:286]: All methods below seem to be class-agonostic (staticmethods).
Maybe we should lift the out and put them in a file called `jax_util.py`?

File: examples/jax/jax_util/__init__.py [hunk 5]
```diff
@@ -0,0 +1,2 @@
+from .datasets import mnist, Dataloader  # noqa: F401
```
Comment (zhisbug) [inline examples/jax/jax_util/__init__.py:1]: Use absolute import instead relative import

**Review by zhisbug — CHANGES_REQUESTED** [review]

File: distml/operator/jax_operator.py (hunk 4 above)
Comment (Ezra-H) [reply distml/operator/jax_operator.py:355]: These methods will be called in replicas/actors.  If put these in `jax_util.py`, the replicas/actors need to realize the operator belongs to which ml framework. That's not what we want.  But we can use `@staticmethod` to decorate these methods.

Comment (zhisbug) [reply distml/operator/jax_operator.py]: I don't know how to deal with these functions now. I'm going to remove them as they are not used by the AllReduceStrategy. Let's revisit them in the next PS pull request.

**Review by zhisbug — APPROVED** [review]

**Pull request merged**